src/System.Net.Sockets/src/System/Net/Sockets/SocketAsyncEventArgs.Windows.cs

```diff
@@ -409,10 +409,17 @@ private unsafe void InnerStartOperationReceiveMessageFrom()
             // WSAMsg also contains a single WSABuffer describing a control buffer.
             PinSocketAddressBuffer();
 
-            // Create and pin a WSAMessageBuffer if none already.
+            // Create a WSAMessageBuffer if none exists yet, and ensure the buffer is pinned.
+            Debug.Assert(
+                !_wsaMessageBufferGCHandle.IsAllocated ||
+                (_wsaMessageBufferGCHandle.Target == _wsaMessageBuffer && _wsaMessageBuffer != null));
+            Debug.Assert((_ptrWSAMessageBuffer != IntPtr.Zero) == _wsaMessageBufferGCHandle.IsAllocated);
```

Comment:
> That's correct, right?

Yes.

> BTW, at some point it would be nice to clean some of this stuff up

Yes, I was thinking the same thing yesterday, in particular because it'd be nice to remove the extra IntPtr field(s) from SAEA.  I'm not doing that in this PR, though.

> we could just allocate the WSAMessageBuffer (and WSABuffer also) on the stack when we do the actual native socket call

For an async call?